{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Let's run intron clustering to annotate alternative splicing events given observed junctions in our cells "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The junctions are loaded from the following path: /gpfs/commons/home/kisaev/LeafletSC/data/raw/junctions/\n",
      "The files in the path are: ['B107926_O8_Blue_Blood_S250.homo.gencode.v30.ERCC.chrM.juncswbarcodes', 'B107925_B5_S284.homo.gencode.v30.ERCC.chrM.juncswbarcodes']\n"
     ]
    }
   ],
   "source": [
    "# Load LeafletSC \n",
    "import LeafletSC\n",
    "import os\n",
    "import pandas as pd \n",
    "\n",
    "# Define path that contains some junction files (only 2 files are used for this example, corresponding to 2 individual cells)\n",
    "juncs_path = \"/gpfs/commons/home/kisaev/LeafletSC/data/raw/junctions/\"\n",
    "print(\"The junctions are loaded from the following path: \" + juncs_path) \n",
    "\n",
    "# print the files in the path \n",
    "print(\"The files in the path are: \" + str(os.listdir(juncs_path)))\n",
    "\n",
    "# define path for saving the output data \n",
    "output_path = \"/gpfs/commons/home/kisaev/LeafletSC/data/processed/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Let's first define some parameters for the analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from LeafletSC.clustering.obtain_intron_clusters import main as obtain_intron_clusters\n",
    "from LeafletSC.clustering.prep_model_input import main as prep_model_input\n",
    "\n",
    "# junc_files defines a path for where junction files can be found, in this case, the path is defined above\n",
    "junc_files = juncs_path\n",
    "\n",
    "# we provide a gtf file for the human genome as well to make better sense of the junctions that are detected in cells\n",
    "# please replace with the path to the gtf file on your system\n",
    "gtf_file=\"/gpfs/commons/groups/knowles_lab/Karin/genome_files/gencode.v43.basic.annotation.gtf\" \n",
    "\n",
    "# define additional parameters \n",
    "sequencing_type = \"single_cell\"\n",
    "\n",
    "# ensure output files are to be saved in output_path \n",
    "output_file = output_path + \"test_intron_clusters\"\n",
    "junc_bed_file= output_path + \"test_juncs.bed\" # you can load this file into IGV to visualize the junction coordinates \n",
    "min_intron_length = 50\n",
    "max_intron_length = 500000\n",
    "threshold_inc = 0.05 \n",
    "min_junc_reads = 2\n",
    "min_num_cells_wjunc = 2\n",
    "keep_singletons = False # ignore junctions that do not share splice sites with any other junction (likely const)\n",
    "junc_suffix = \"*.juncswbarcodes\" # depends on how you ran regtools "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run intron clustering "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading in junction files from /gpfs/commons/home/kisaev/LeafletSC/data/raw/junctions\n",
      "The number of regtools junction files to be processed is 2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 2/2 [00:00<00:00, 12.48it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total number of files that could not be read is 0 as these had no junctions\n",
      "The gtf file you provided is /gpfs/commons/groups/knowles_lab/Karin/genome_files/gencode.v43.basic.annotation.gtf\n",
      "Now reading gtf file using gtfparse\n",
      "This step may take a while depending on the size of your gtf file\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'level', 'tag', 'transcript_id', 'transcript_type', 'transcript_name', 'transcript_support_level', 'havana_transcript', 'exon_number', 'exon_id', 'hgnc_id', 'havana_gene', 'ont', 'protein_id', 'ccdsid', 'artif_dupl']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading gtf file took 45.91 seconds\n",
      "The number of unique exons is 411865\n",
      "The number of unique transcript ids is 115526\n",
      "The number of unique gene ids is 62668\n",
      "Done extracting exons from gtf file\n",
      "Loading files obtained from single_cell sequencing\n",
      "Filtering based on intron length\n",
      "Cleaning up 'chrom' column\n",
      "Making gr object from all junctions across all cell types\n",
      "The summary statistics for counts_total across junctions are: \n",
      "count    14417.000000\n",
      "mean       167.274745\n",
      "std        496.475133\n",
      "min          1.000000\n",
      "25%         15.000000\n",
      "50%         70.000000\n",
      "75%        172.000000\n",
      "max      17913.000000\n",
      "Name: counts_total, dtype: float64\n",
      "The number of junctions with only 1 read across the board is: \n",
      "1007\n",
      "The number of junctions after filtering for minimum junction reads is 12876\n",
      "The number of junctions prior to assessing distance to exons is 12876\n",
      "Annotating junctions with known exons based on input gtf file\n",
      "The number of junctions after assessing distance to exons is 11813\n",
      "Clustering intron splicing events by gene_id\n",
      "The number of clusters after clustering by gene_id is 11253\n",
      "1     11007\n",
      "3       333\n",
      "2       180\n",
      "4        80\n",
      "5        35\n",
      "7        28\n",
      "8        24\n",
      "22       22\n",
      "19       19\n",
      "6        18\n",
      "9        18\n",
      "14       14\n",
      "13       13\n",
      "12       12\n",
      "10       10\n",
      "Name: Count, dtype: int64\n",
      "The number of clusters after removing singletons is 246\n",
      "Confirming that junctions in each cluster share splice sites\n",
      "The number of clusters after filtering for shared splice sites is 237\n",
      "3     351\n",
      "2     198\n",
      "6      36\n",
      "5      30\n",
      "18     18\n",
      "8      16\n",
      "7      14\n",
      "13     13\n",
      "10     10\n",
      "4       8\n",
      "Name: Count, dtype: int64\n",
      "The number of clusters after removing singletons is 237\n",
      "The number of clusters to be finally evaluated is 237\n",
      "The number of junctions to be finally evaluated is 694\n",
      "You can find the output file here: /gpfs/commons/home/kisaev/LeafletSC/data/processed/test_intron_clusters_50_500000_2_20240309_single_cell.gz\n",
      "Finished obtaining intron cluster files!\n"
     ]
    }
   ],
   "source": [
    "obtain_intron_clusters(junc_files=junc_files, gtf_file=gtf_file, output_file=output_file, \n",
    "                       sequencing_type=sequencing_type, junc_bed_file=junc_bed_file, \n",
    "                       threshold_inc=threshold_inc, min_intron = min_intron_length,\n",
    "                       max_intron=max_intron_length, min_junc_reads=min_junc_reads,\n",
    "                       singleton=keep_singletons, strict_filter=False, \n",
    "                       junc_suffix=junc_suffix, min_num_cells_wjunc=min_num_cells_wjunc,\n",
    "                       filter_low_juncratios_inclust=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Now let's convert the intron clusters to a format that can be used by LeafletSC )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "intron_clusters = \"/gpfs/commons/home/kisaev/LeafletSC/data/processed/test_intron_clusters_50_500000_2_20240309_single_cell.gz\" # path to the intron clusters file\n",
    "output_file = output_path + \"test_model_input\" # name of the output file\n",
    "has_genes = \"yes\" # since we used a gtf file to obtain the intron clusters, we can set this to yes\n",
    "chunk_size = 5000 # number of junctions to process at a time from the intron clusters files\n",
    "metadata = None # can replace with path, if metadata is available for cells (cell type, origin, library ID...)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The intron clusters file you provided is /gpfs/commons/home/kisaev/LeafletSC/data/processed/test_intron_clusters_50_500000_2_20240309_single_cell.gz, reading in chunks of 5000 lines\n",
      "Processing intron clusters\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1it [00:00, 52.99it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done processing intron clusters\n",
      "The number of intron clusters evaluated is 237\n",
      "The number of junctions evaluated is 694\n",
      "The number of cells evaluated is 2\n",
      "The number of clusters with only one junction is 0\n",
      "A gtf file was used to generate intron clusters\n",
      "The number of total cells evaluated is 2\n",
      "The number of cells per cell type is:\n",
      "cell_type\n",
      "/gpfs/commons/home/kisaev/LeafletSC/data/raw/junctions/B107925_B5_S284.homo.gencode.v30.ERCC.chrM.juncswbarcodes               1\n",
      "/gpfs/commons/home/kisaev/LeafletSC/data/raw/junctions/B107926_O8_Blue_Blood_S250.homo.gencode.v30.ERCC.chrM.juncswbarcodes    1\n",
      "Name: cell_type, dtype: int64\n",
      "['B107925_B5_S284.homo.gencode.v30.ERCC.chrM_/gpfs/commons/home/kisaev/LeafletSC/data/raw/junctions/B107925_B5_S284.homo.gencode.v30.ERCC.chrM.juncswbarcodes'\n",
      " 'B107926_O8_Blue_Blood_S250.homo.gencode.v30.ERCC.chrM_/gpfs/commons/home/kisaev/LeafletSC/data/raw/junctions/B107926_O8_Blue_Blood_S250.homo.gencode.v30.ERCC.chrM.juncswbarcodes']\n",
      "You can find the resulting file at /gpfs/commons/home/kisaev/LeafletSC/data/processed/test_model_input.h5\n",
      "Done generating input file for Leaflet model. This process took 0 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "prep_model_input.main(intron_clusters, output_file, has_genes, chunk_size, metadata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Take a quick look at the input file that will go into the model to get familiarized with all the columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                             cell_id  Cluster  Cluster_Counts  \\\n",
      "0  B107925_B5_S284.homo.gencode.v30.ERCC.chrM_/gp...       70             153   \n",
      "1  B107925_B5_S284.homo.gencode.v30.ERCC.chrM_/gp...       70             153   \n",
      "2  B107925_B5_S284.homo.gencode.v30.ERCC.chrM_/gp...      151             376   \n",
      "3  B107925_B5_S284.homo.gencode.v30.ERCC.chrM_/gp...      151             376   \n",
      "4  B107925_B5_S284.homo.gencode.v30.ERCC.chrM_/gp...      151             376   \n",
      "\n",
      "             junction_id             gene_id  junc_count  \\\n",
      "0  1_175004814_175006744  ENSG00000116161.18           8   \n",
      "1  1_175004833_175006744  ENSG00000116161.18         145   \n",
      "2  1_145608175_145616233  ENSG00000117262.19         218   \n",
      "3  1_145608175_145618319  ENSG00000117262.19          20   \n",
      "4  1_145616293_145618319  ENSG00000117262.19         138   \n",
      "\n",
      "                                           cell_type  junc_ratio  \\\n",
      "0  /gpfs/commons/home/kisaev/LeafletSC/data/raw/j...    0.052288   \n",
      "1  /gpfs/commons/home/kisaev/LeafletSC/data/raw/j...    0.947712   \n",
      "2  /gpfs/commons/home/kisaev/LeafletSC/data/raw/j...    0.579787   \n",
      "3  /gpfs/commons/home/kisaev/LeafletSC/data/raw/j...    0.053191   \n",
      "4  /gpfs/commons/home/kisaev/LeafletSC/data/raw/j...    0.367021   \n",
      "\n",
      "   cell_id_index  junction_id_index  \n",
      "0              0                386  \n",
      "1              0                387  \n",
      "2              0                332  \n",
      "3              0                333  \n",
      "4              0                334  \n"
     ]
    }
   ],
   "source": [
    "model_input_data = \"/gpfs/commons/home/kisaev/LeafletSC/data/processed/test_model_input.h5\"\n",
    "summarized_data = pd.read_hdf(model_input_data, 'df')\n",
    "print(summarized_data.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Note that fow now, the values in cell_type default to the cell's path, in the future it will be possible to specify the cell type in the metadata file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['/gpfs/commons/home/kisaev/LeafletSC/data/raw/junctions/B107925_B5_S284.homo.gencode.v30.ERCC.chrM.juncswbarcodes',\n",
       "       '/gpfs/commons/home/kisaev/LeafletSC/data/raw/junctions/B107926_O8_Blue_Blood_S250.homo.gencode.v30.ERCC.chrM.juncswbarcodes'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summarized_data.cell_type.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['cell_id', 'Cluster', 'Cluster_Counts', 'junction_id', 'gene_id',\n",
      "       'junc_count', 'cell_type', 'junc_ratio', 'cell_id_index',\n",
      "       'junction_id_index'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "# let's see all the columns in the summarized data\n",
    "print(summarized_data.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### We can quickly visualize the overall junction usage ratio distribution across all cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Axes(0.125,0.11;0.775x0.77)\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAigAAAGdCAYAAAA44ojeAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAgCklEQVR4nO3de3CU9fm/8XeOG6IsMWJONSBgERQUGgpEsR4IhMOgVmYUZRh0KLQanJFMPeCBoy2UYdSRiTJaBTsD0tpRWpEGIhQpGlAjTBWQysGihYQCDQFSliX5/P74TvbXENBs3H323nC9ZjJjnn3y5N6bTbjcJCTBOecEAABgSGKsBwAAADgbgQIAAMwhUAAAgDkECgAAMIdAAQAA5hAoAADAHAIFAACYQ6AAAABzkmM9QFs0NjbqwIED6tixoxISEmI9DgAAaAXnnI4fP668vDwlJn77cyRxGSgHDhxQfn5+rMcAAABt8PXXX+vyyy//1nPiMlA6duwo6f/uoN/vj9h1g8Gg1q5dq+HDhyslJSVi10Vz7Nk77Nob7Nkb7Nk70dp1XV2d8vPzQ3+Pf5u4DJSmL+v4/f6IB0p6err8fj8P/ihiz95h195gz95gz96J9q5b8+0ZfJMsAAAwh0ABAADmECgAAMAcAgUAAJhDoAAAAHMIFAAAYA6BAgAAzCFQAACAOQQKAAAwh0ABAADmECgAAMAcAgUAAJhDoAAAAHMIFAAAYE5yrAewqM+sNQo0fPevgrbkq/mjYz0CAAARwzMoAADAHAIFAACYQ6AAAABzCBQAAGAOgQIAAMwhUAAAgDkECgAAMIdAAQAA5hAoAADAHAIFAACYQ6AAAABzCBQAAGAOgQIAAMwhUAAAgDkECgAAMIdAAQAA5hAoAADAHAIFAACYQ6AAAABzCBQAAGAOgQIAAMwhUAAAgDkECgAAMIdAAQAA5hAoAADAHAIFAACYQ6AAAABzCBQAAGAOgQIAAMwhUAAAgDkECgAAMIdAAQAA5hAoAADAHAIFAACYQ6AAAABzCBQAAGAOgQIAAMwhUAAAgDkECgAAMIdAAQAA5hAoAADAHAIFAACYQ6AAAABzCBQAAGAOgQIAAMwhUAAAgDlhBcq8efP04x//WB07dlRWVpbuuOMO7dq1q9k5p06dUklJiS699FJdfPHFGjt2rGpqapqds3//fo0ePVrp6enKysrSI488ojNnznz/ewMAANqFsALl/fffV0lJiTZv3qyKigoFg0ENHz5cJ0+eDJ0zbdo0vfPOO3rzzTf1/vvv68CBA7rzzjtDtzc0NGj06NE6ffq0PvzwQ73++utaunSpZsyYEbl7BQAA4lpyOCeXl5c3e33p0qXKyspSVVWVfvKTn+jYsWN69dVXtXz5ct16662SpCVLlqh3797avHmzBg8erLVr12rHjh167733lJ2drX79+mnu3Ll67LHHNGvWLKWmpkbu3gEAgLgUVqCc7dixY5KkzMxMSVJVVZWCwaCKiopC5/Tq1UtdunRRZWWlBg8erMrKSvXt21fZ2dmhc4qLi/XAAw9o+/bt6t+/f4v3EwgEFAgEQq/X1dVJkoLBoILB4Pe5C800XcuX6CJ2Ta9Ecg/R1jRrPM0cr9i1N9izN9izd6K163Cu1+ZAaWxs1MMPP6wbbrhBffr0kSRVV1crNTVVGRkZzc7Nzs5WdXV16Jz/jZOm25tuO5d58+Zp9uzZLY6vXbtW6enpbb0L5zV3QGPErxltq1evjvUIYauoqIj1CBcMdu0N9uwN9uydSO+6vr6+1ee2OVBKSkr0+eefa9OmTW29RKtNnz5dpaWlodfr6uqUn5+v4cOHy+/3R+z9BINBVVRU6OlPEhVoTIjYdb3w+aziWI/Qak17HjZsmFJSUmI9TrvGrr3Bnr3Bnr0TrV03fQWkNdoUKFOnTtWqVau0ceNGXX755aHjOTk5On36tGpra5s9i1JTU6OcnJzQOR999FGz6zX9lE/TOWfz+Xzy+XwtjqekpETlQRpoTFCgIb4CJR4/WKP154eW2LU32LM32LN3Ir3rcK4V1k/xOOc0depUvf3221q/fr26devW7PaCggKlpKRo3bp1oWO7du3S/v37VVhYKEkqLCzUZ599pkOHDoXOqaiokN/v19VXXx3OOAAAoJ0K6xmUkpISLV++XH/605/UsWPH0PeMdOrUSR06dFCnTp00adIklZaWKjMzU36/Xw899JAKCws1ePBgSdLw4cN19dVXa8KECVqwYIGqq6v11FNPqaSk5JzPkgAAgAtPWIHy0ksvSZJuvvnmZseXLFmi++67T5L03HPPKTExUWPHjlUgEFBxcbFefPHF0LlJSUlatWqVHnjgARUWFuqiiy7SxIkTNWfOnO93TwAAQLsRVqA4990/fpuWlqaysjKVlZWd95yuXbvG5U+dAAAAb/C7eAAAgDkECgAAMIdAAQAA5hAoAADAHAIFAACYQ6AAAABzCBQAAGAOgQIAAMwhUAAAgDkECgAAMIdAAQAA5hAoAADAHAIFAACYQ6AAAABzCBQAAGAOgQIAAMwhUAAAgDkECgAAMIdAAQAA5hAoAADAHAIFAACYQ6AAAABzCBQAAGAOgQIAAMwhUAAAgDkECgAAMIdAAQAA5hAoAADAHAIFAACYQ6AAAABzCBQAAGAOgQIAAMwhUAAAgDkECgAAMIdAAQAA5hAoAADAHAIFAACYQ6AAAABzCBQAAGAOgQIAAMwhUAAAgDkECgAAMIdAAQAA5hAoAADAHAIFAACYQ6AAAABzCBQAAGAOgQIAAMwhUAAAgDkECgAAMIdAAQAA5hAoAADAHAIFAACYQ6AAAABzCBQAAGAOgQIAAMwhUAAAgDkECgAAMIdAAQAA5hAoAADAHAIFAACYQ6AAAABzCBQAAGAOgQIAAMwJO1A2btyoMWPGKC8vTwkJCVq5cmWz2++77z4lJCQ0exkxYkSzc44eParx48fL7/crIyNDkyZN0okTJ77XHQEAAO1H2IFy8uRJXXfddSorKzvvOSNGjNDBgwdDL2+88Uaz28ePH6/t27eroqJCq1at0saNGzVlypTwpwcAAO1ScrhvMHLkSI0cOfJbz/H5fMrJyTnnbTt37lR5ebk+/vhjDRgwQJK0aNEijRo1SgsXLlReXl64IwEAgHYmKt+DsmHDBmVlZemqq67SAw88oCNHjoRuq6ysVEZGRihOJKmoqEiJiYnasmVLNMYBAABxJuxnUL7LiBEjdOedd6pbt27as2ePnnjiCY0cOVKVlZVKSkpSdXW1srKymg+RnKzMzExVV1ef85qBQECBQCD0el1dnSQpGAwqGAxGbPama/kSXcSu6ZVI7iHammaNp5njFbv2Bnv2Bnv2TrR2Hc71Ih4o48aNC/133759de2116pHjx7asGGDhg4d2qZrzps3T7Nnz25xfO3atUpPT2/zrOczd0BjxK8ZbatXr471CGGrqKiI9QgXDHbtDfbsDfbsnUjvur6+vtXnRjxQzta9e3d17txZu3fv1tChQ5WTk6NDhw41O+fMmTM6evToeb9vZfr06SotLQ29XldXp/z8fA0fPlx+vz9iswaDQVVUVOjpTxIVaEyI2HW98Pms4liP0GpNex42bJhSUlJiPU67xq69wZ69wZ69E61dN30FpDWiHijffPONjhw5otzcXElSYWGhamtrVVVVpYKCAknS+vXr1djYqEGDBp3zGj6fTz6fr8XxlJSUqDxIA40JCjTEV6DE4wdrtP780BK79gZ79gZ79k6kdx3OtcIOlBMnTmj37t2h1/ft26dt27YpMzNTmZmZmj17tsaOHaucnBzt2bNHjz76qK688koVF//f/+H37t1bI0aM0OTJk7V48WIFg0FNnTpV48aN4yd4AACApDb8FM8nn3yi/v37q3///pKk0tJS9e/fXzNmzFBSUpL+/ve/67bbblPPnj01adIkFRQU6G9/+1uzZ0CWLVumXr16aejQoRo1apSGDBmil19+OXL3CgAAxLWwn0G5+eab5dz5f8plzZo133mNzMxMLV++PNx3DQAALhD8Lh4AAGAOgQIAAMwhUAAAgDkECgAAMIdAAQAA5hAoAADAHAIFAACYQ6AAAABzCBQAAGAOgQIAAMwhUAAAgDkECgAAMIdAAQAA5hAoAADAHAIFAACYQ6AAAABzCBQAAGAOgQIAAMwhUAAAgDkECgAAMIdAAQAA5hAoAADAHAIFAACYQ6AAAABzCBQAAGAOgQIAAMwhUAAAgDkECgAAMIdAAQAA5hAoAADAHAIFAACYQ6AAAABzCBQAAGBOcqwHABB9fWatUaAhIdZjtNpX80fHegQAMcYzKAAAwBwCBQAAmEOgAAAAcwgUAABgDoECAADMIVAAAIA5BAoAADCHQAEAAOYQKAAAwBwCBQAAmEOgAAAAcwgUAABgDoECAADMIVAAAIA5BAoAADCHQAEAAOYQKAAAwBwCBQAAmEOgAAAAcwgUAABgDoECAADMSY71AABwtisefzfWI4TFl+S0YGCsp4BV8fZ4lmw8pnkGBQAAmEOgAAAAcwgUAABgDoECAADMIVAAAIA5BAoAADCHQAEAAOYQKAAAwBwCBQAAmBN2oGzcuFFjxoxRXl6eEhIStHLlyma3O+c0Y8YM5ebmqkOHDioqKtKXX37Z7JyjR49q/Pjx8vv9ysjI0KRJk3TixInvdUcAAED7EXagnDx5Utddd53KysrOefuCBQv0wgsvaPHixdqyZYsuuugiFRcX69SpU6Fzxo8fr+3bt6uiokKrVq3Sxo0bNWXKlLbfCwAA0K6E/bt4Ro4cqZEjR57zNuecnn/+eT311FO6/fbbJUm/+93vlJ2drZUrV2rcuHHauXOnysvL9fHHH2vAgAGSpEWLFmnUqFFauHCh8vLyvsfdAQAA7UFEf1ngvn37VF1draKiotCxTp06adCgQaqsrNS4ceNUWVmpjIyMUJxIUlFRkRITE7Vlyxb99Kc/bXHdQCCgQCAQer2urk6SFAwGFQwGIzZ/07V8iS5i1/RKJPcQbU2zxtPM8SqeH9PxpGm/PKajK14/d/iS4u/jL1qP6XCuF9FAqa6uliRlZ2c3O56dnR26rbq6WllZWc2HSE5WZmZm6JyzzZs3T7Nnz25xfO3atUpPT4/E6M3MHdAY8WtG2+rVq2M9QtgqKipiPcIFIx4f0/GIx7Q34m3Psf6twN9HpHddX1/f6nMjGijRMn36dJWWloZer6urU35+voYPHy6/3x+x9xMMBlVRUaGnP0lUoDEhYtf1wuezimM9Qqs17XnYsGFKSUmJ9TjtWjw/puOJL9Fp7oBGHtNRFq+fO/rMWhPrEcIWrcd001dAWiOigZKTkyNJqqmpUW5ubuh4TU2N+vXrFzrn0KFDzd7uzJkzOnr0aOjtz+bz+eTz+VocT0lJicqDNNCYoEBDfH0yj6cP1ibR+vNDS/H4mI5HPKa9EW97juePvUjvOpxrRfTfQenWrZtycnK0bt260LG6ujpt2bJFhYWFkqTCwkLV1taqqqoqdM769evV2NioQYMGRXIcAAAQp8J+BuXEiRPavXt36PV9+/Zp27ZtyszMVJcuXfTwww/rmWee0Q9/+EN169ZNTz/9tPLy8nTHHXdIknr37q0RI0Zo8uTJWrx4sYLBoKZOnapx48bxEzwAAEBSGwLlk08+0S233BJ6vel7QyZOnKilS5fq0Ucf1cmTJzVlyhTV1tZqyJAhKi8vV1paWuhtli1bpqlTp2ro0KFKTEzU2LFj9cILL0Tg7gAAgPYg7EC5+eab5dz5f2QqISFBc+bM0Zw5c857TmZmppYvXx7uuwYAABcIfhcPAAAwh0ABAADmECgAAMAcAgUAAJhDoAAAAHMIFAAAYA6BAgAAzCFQAACAOQQKAAAwh0ABAADmECgAAMAcAgUAAJhDoAAAAHMIFAAAYA6BAgAAzEmO9QBAPLni8XdjPUJYfElOCwbGegoACB/PoAAAAHMIFAAAYA6BAgAAzCFQAACAOQQKAAAwh0ABAADmECgAAMAcAgUAAJhDoAAAAHMIFAAAYA6BAgAAzCFQAACAOQQKAAAwh0ABAADmECgAAMAcAgUAAJhDoAAAAHMIFAAAYA6BAgAAzCFQAACAOQQKAAAwh0ABAADmECgAAMAcAgUAAJhDoAAAAHMIFAAAYA6BAgAAzCFQAACAOQQKAAAwh0ABAADmECgAAMAcAgUAAJhDoAAAAHMIFAAAYA6BAgAAzCFQAACAOcmxHgCRccXj78Z6hFbzJTktGBjrKQAAlvEMCgAAMIdAAQAA5hAoAADAHAIFAACYQ6AAAABzCBQAAGAOgQIAAMwhUAAAgDkECgAAMIdAAQAA5hAoAADAnIgHyqxZs5SQkNDspVevXqHbT506pZKSEl166aW6+OKLNXbsWNXU1ER6DAAAEMei8gzKNddco4MHD4ZeNm3aFLpt2rRpeuedd/Tmm2/q/fff14EDB3TnnXdGYwwAABCnovLbjJOTk5WTk9Pi+LFjx/Tqq69q+fLluvXWWyVJS5YsUe/evbV582YNHjw4GuMAAIA4E5VA+fLLL5WXl6e0tDQVFhZq3rx56tKli6qqqhQMBlVUVBQ6t1evXurSpYsqKyvPGyiBQECBQCD0el1dnSQpGAwqGAxGbO6ma/kSXcSuiZaa9lswp1yBxoQYTxMeX1KsJwhP0655TEdX034j+fkILTXtN9727EuKv4+/aD2mw7legnMuopv7y1/+ohMnTuiqq67SwYMHNXv2bP3rX//S559/rnfeeUf3339/s9iQpIEDB+qWW27Rb37zm3Nec9asWZo9e3aL48uXL1d6enokxwcAAFFSX1+ve++9V8eOHZPf7//WcyMeKGerra1V165d9eyzz6pDhw5tCpRzPYOSn5+vw4cPf+cdDEcwGFRFRYWe/iQx7v7PPp74Ep3mDmhkzx5g195o2vOwYcOUkpIS63HarabP0fG25z6z1sR6hLBF6zFdV1enzp07typQovIlnv+VkZGhnj17avfu3Ro2bJhOnz6t2tpaZWRkhM6pqak55/esNPH5fPL5fC2Op6SkROVBGmhMUKCBT+bRxp69w669Ea3PSWgu3vYczx97kd51ONeK+r+DcuLECe3Zs0e5ubkqKChQSkqK1q1bF7p9165d2r9/vwoLC6M9CgAAiBMRfwbll7/8pcaMGaOuXbvqwIEDmjlzppKSknTPPfeoU6dOmjRpkkpLS5WZmSm/36+HHnpIhYWF/AQPAAAIiXigfPPNN7rnnnt05MgRXXbZZRoyZIg2b96syy67TJL03HPPKTExUWPHjlUgEFBxcbFefPHFSI8BAADiWMQDZcWKFd96e1pamsrKylRWVhbpdw0AANoJfhcPAAAwh0ABAADmECgAAMAcAgUAAJhDoAAAAHMIFAAAYA6BAgAAzCFQAACAOQQKAAAwh0ABAADmECgAAMAcAgUAAJhDoAAAAHMIFAAAYA6BAgAAzCFQAACAOQQKAAAwh0ABAADmECgAAMAcAgUAAJhDoAAAAHMIFAAAYA6BAgAAzCFQAACAOQQKAAAwh0ABAADmJMd6AABoL/rMWqNAQ0Ksx2i1r+aPjvUIwHnxDAoAADCHQAEAAOYQKAAAwBwCBQAAmEOgAAAAcwgUAABgDoECAADMIVAAAIA5BAoAADCHQAEAAOYQKAAAwBx+Fw8AXKCuePzdWI8QFl+S04KBsZ4CXuEZFAAAYA6BAgAAzCFQAACAOQQKAAAwh0ABAADmECgAAMAcAgUAAJjDv4MCAIgrfWatUaAhIdZjIMp4BgUAAJhDoAAAAHMIFAAAYA6BAgAAzCFQAACAOQQKAAAwh0ABAADmECgAAMAcAgUAAJhDoAAAAHMIFAAAYA6BAgAAzCFQAACAOQQKAAAwh0ABAADmECgAAMAcAgUAAJgT00ApKyvTFVdcobS0NA0aNEgfffRRLMcBAABGxCxQfv/736u0tFQzZ87Up59+quuuu07FxcU6dOhQrEYCAABGxCxQnn32WU2ePFn333+/rr76ai1evFjp6el67bXXYjUSAAAwIjkW7/T06dOqqqrS9OnTQ8cSExNVVFSkysrKFucHAgEFAoHQ68eOHZMkHT16VMFgMGJzBYNB1dfXKzmYqIbGhIhdF80lNzrV1zeyZw+wa2+wZ2+wZ+807frIkSNKSUmJ2HWPHz8uSXLOffcMEXuvYTh8+LAaGhqUnZ3d7Hh2dra++OKLFufPmzdPs2fPbnG8W7duUZsR0XVvrAe4gLBrb7Bnb7Bn70Rz18ePH1enTp2+9ZyYBEq4pk+frtLS0tDrjY2NOnr0qC699FIlJESuouvq6pSfn6+vv/5afr8/YtdFc+zZO+zaG+zZG+zZO9HatXNOx48fV15e3neeG5NA6dy5s5KSklRTU9PseE1NjXJyclqc7/P55PP5mh3LyMiI2nx+v58HvwfYs3fYtTfYszfYs3eisevveuakSUy+STY1NVUFBQVat25d6FhjY6PWrVunwsLCWIwEAAAMidmXeEpLSzVx4kQNGDBAAwcO1PPPP6+TJ0/q/vvvj9VIAADAiJgFyt13361///vfmjFjhqqrq9WvXz+Vl5e3+MZZL/l8Ps2cObPFl5MQWezZO+zaG+zZG+zZOxZ2neBa87M+AAAAHuJ38QAAAHMIFAAAYA6BAgAAzCFQAACAORdcoJSVlemKK65QWlqaBg0apI8++uhbz3/zzTfVq1cvpaWlqW/fvlq9erVHk8a3cPb8yiuv6MYbb9Qll1yiSy65REVFRd/554L/L9zHdJMVK1YoISFBd9xxR3QHbCfC3XNtba1KSkqUm5srn8+nnj178vmjFcLd8/PPP6+rrrpKHTp0UH5+vqZNm6ZTp055NG182rhxo8aMGaO8vDwlJCRo5cqV3/k2GzZs0I9+9CP5fD5deeWVWrp0adTnlLuArFixwqWmprrXXnvNbd++3U2ePNllZGS4mpqac57/wQcfuKSkJLdgwQK3Y8cO99RTT7mUlBT32WefeTx5fAl3z/fee68rKytzW7dudTt37nT33Xef69Spk/vmm288njz+hLvrJvv27XM/+MEP3I033uhuv/12b4aNY+HuORAIuAEDBrhRo0a5TZs2uX379rkNGza4bdu2eTx5fAl3z8uWLXM+n88tW7bM7du3z61Zs8bl5ua6adOmeTx5fFm9erV78skn3VtvveUkubfffvtbz9+7d69LT093paWlbseOHW7RokUuKSnJlZeXR3XOCypQBg4c6EpKSkKvNzQ0uLy8PDdv3rxznn/XXXe50aNHNzs2aNAg9/Of/zyqc8a7cPd8tjNnzriOHTu6119/PVojthtt2fWZM2fc9ddf737729+6iRMnEiitEO6eX3rpJde9e3d3+vRpr0ZsF8Ldc0lJibv11lubHSstLXU33HBDVOdsT1oTKI8++qi75pprmh27++67XXFxcRQnc+6C+RLP6dOnVVVVpaKiotCxxMREFRUVqbKy8pxvU1lZ2ex8SSouLj7v+Wjbns9WX1+vYDCozMzMaI3ZLrR113PmzFFWVpYmTZrkxZhxry17/vOf/6zCwkKVlJQoOztbffr00a9//Ws1NDR4NXbcacuer7/+elVVVYW+DLR3716tXr1ao0aN8mTmC0Ws/i6Mi99mHAmHDx9WQ0NDi3+pNjs7W1988cU536a6uvqc51dXV0dtznjXlj2f7bHHHlNeXl6LDwg015Zdb9q0Sa+++qq2bdvmwYTtQ1v2vHfvXq1fv17jx4/X6tWrtXv3bj344IMKBoOaOXOmF2PHnbbs+d5779Xhw4c1ZMgQOed05swZ/eIXv9ATTzzhxcgXjPP9XVhXV6f//ve/6tChQ1Te7wXzDAriw/z587VixQq9/fbbSktLi/U47crx48c1YcIEvfLKK+rcuXOsx2nXGhsblZWVpZdfflkFBQW6++679eSTT2rx4sWxHq1d2bBhg37961/rxRdf1Keffqq33npL7777rubOnRvr0RABF8wzKJ07d1ZSUpJqamqaHa+pqVFOTs453yYnJyes89G2PTdZuHCh5s+fr/fee0/XXnttNMdsF8Ld9Z49e/TVV19pzJgxoWONjY2SpOTkZO3atUs9evSI7tBxqC2P6dzcXKWkpCgpKSl0rHfv3qqurtbp06eVmpoa1ZnjUVv2/PTTT2vChAn62c9+Jknq27evTp48qSlTpujJJ59UYiL/Dx4J5/u70O/3R+3ZE+kCegYlNTVVBQUFWrduXehYY2Oj1q1bp8LCwnO+TWFhYbPzJamiouK856Nte5akBQsWaO7cuSovL9eAAQO8GDXuhbvrXr166bPPPtO2bdtCL7fddptuueUWbdu2Tfn5+V6OHzfa8pi+4YYbtHv37lAAStI//vEP5ebmEifn0ZY919fXt4iQpih0/Jq5iInZ34VR/RZcY1asWOF8Pp9bunSp27Fjh5syZYrLyMhw1dXVzjnnJkyY4B5//PHQ+R988IFLTk52CxcudDt37nQzZ87kx4xbIdw9z58/36Wmpro//vGP7uDBg6GX48ePx+ouxI1wd302foqndcLd8/79+13Hjh3d1KlT3a5du9yqVatcVlaWe+aZZ2J1F+JCuHueOXOm69ixo3vjjTfc3r173dq1a12PHj3cXXfdFau7EBeOHz/utm7d6rZu3eokuWeffdZt3brV/fOf/3TOOff444+7CRMmhM5v+jHjRx55xO3cudOVlZXxY8bRsGjRItelSxeXmprqBg4c6DZv3hy67aabbnITJ05sdv4f/vAH17NnT5eamuquueYa9+6773o8cXwKZ89du3Z1klq8zJw50/vB41C4j+n/RaC0Xrh7/vDDD92gQYOcz+dz3bt3d7/61a/cmTNnPJ46/oSz52Aw6GbNmuV69Ojh0tLSXH5+vnvwwQfdf/7zH+8HjyN//etfz/k5t2m3EydOdDfddFOLt+nXr59LTU113bt3d0uWLIn6nAnO8TwYAACw5YL5HhQAABA/CBQAAGAOgQIAAMwhUAAAgDkECgAAMIdAAQAA5hAoAADAHAIFAACYQ6AAAABzCBQAAGAOgQIAAMwhUAAAgDn/Dzeeu2s8REJIAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print(summarized_data.junc_ratio.hist())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Now we have everything need to run the Leaflet mixture model! Please refer to the next notebook for the next steps."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "leafcutter-sc",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
